### PR TITLE
Add Invoke-based automation tasks.

### LINKED ISSUE #24: Consider switching from Make to Invoke
Makefiles can get rather complicated, with somewhat arcane and cumbersome syntax. Would we be better off switching to [Invoke](https://github.com/pyinvoke/invoke), which would allow us to reason in Python?

### PR DESCRIPTION
**Pull request by justinmayer**
Fixes #24

### CODE REVIEW

File: CONTRIBUTING.rst [hunk 1]
```diff
@@ -4,8 +4,29 @@ Contributing
...
+Then clone the source for your fork and add the upstream project as a Git remote::
+
+   git clone https://github.com/YOUR_USERNAME/kagi.git
+   cd kagi
+   git remote add upstream https://github.com/justinmayer/kagi.git
+
+The last two steps will install the needed dependencies and set up the project::
+
+   poetry install
+   invoke setup
```
Comment (Natim) [inline CONTRIBUTING.rst]: :+1:

File: tasks.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+TOOLS = ["poetry", "pre-commit"]
+PORT = os.environ.get("SERVER_PORT", 8000)
+
+ACTIVE_VENV = os.environ.get("VIRTUAL_ENV", None)
+VENV_HOME = Path(os.environ.get("WORKON_HOME", "~/.local/share/virtualenvs"))
+VENV_PATH = Path(ACTIVE_VENV) if ACTIVE_VENV else (VENV_HOME / "kagi")
+VENV = str(VENV_PATH.expanduser())
+
+PRECOMMIT = which("pre-commit") if which("pre-commit") else (VENV / "pre-commit")
```
Comment (Natim) [inline tasks.py]:
```suggestion
PRECOMMIT = which("pre-commit") if which("pre-commit") else (VENV / "bin" / "pre-commit")
```

File: tasks.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+POETRY = which("poetry") if which("poetry") else (VENV / "poetry")
```
Comment (Natim) [inline tasks.py]:
```suggestion
POETRY = which("poetry") if which("poetry") else (VENV / "bin" / "poetry")
```

File: tasks.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+
+@task
+def build(c, docs=True):
```
Comment (Natim) [inline tasks.py]:
```suggestion
def build(c, docs=True):
    if docs:
        build_docs(c)

def build_docs(c):
```

Comment (Natim) [reply tasks.py]: To be honest, I am not sure about the build command at all, it seems a bit too generic.

File: tasks.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+@task(build)
+def livedocs(c):
+    """Serve docs at http://localhost:$PORT/ (default port is 8000)"""
+    from livereload import Server
+
+    server = Server()
+    server.watch("docs/conf.py", lambda: build(c))
+    server.watch("CONTRIBUTING.rst", lambda: build(c))
+    server.watch("docs/*.rst", lambda: build(c))
+    server.serve(port=PORT, root="docs/_build")
```
Comment (Natim) [inline tasks.py]: Should we have a `DOC_PORT` and a `DEMO_PORT`?

File: tasks.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+
+@task
+def serve(c):
+    """Serve site at https://localhost:$PORT/ (default port is 8000)"""
+    with cd("testproj"):
+        c.run(
+            f"{VENV}/bin/python manage.py runserver_plus "
```
Comment (Natim) [inline tasks.py]: I realized while writting the troubleshooting, that everything works fine in localhost even if we don't use SSL. I guess would make it easier for people if we relied directly on `runserver`

File: tasks.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+@task
+def migrate(c):
+    with cd("testproj"):
+        c.run(f"{VENV}/bin/python manage.py migrate")
+
+
+@task
+def black(c, diff=False):
+    """Run Black auto-formatter, optionally with --diff"""
+    if diff:
```
Comment (Natim) [inline tasks.py]: we could use check also for the linter.

**Comment by Natim** [comment]
Really promising !!!

**Comment by justinmayer** [comment]
Good suggestions! I'll add a follow-up commit, and you can add another PR afterward in case I miss any. 😅

File: tasks.py (hunk 5 above)
Comment (justinmayer) [reply tasks.py]: I thought about that. I wasn't sure whether anyone would want to have both running at the same time or not. But I suppose we can separate the options and let users decide. :smile:

File: tasks.py (hunk 7 above)
Comment (justinmayer) [reply tasks.py]: Do you mean `--diff` for the `lint` task? Or `--check` for the `black` task?

File: tasks.py (hunk 4 above)
Comment (justinmayer) [reply tasks.py]: Agreed. As of bad1205 those are now `docs` and `viewdocs` tasks.

File: tasks.py (hunk 5 above)
Comment (justinmayer) [reply tasks.py]: Separate `DOCS_PORT` and `DEMO_PORT` variables now exist as of bad1205.

File: tasks.py (hunk 6 above)
Comment (justinmayer) [reply tasks.py]: Agreed. See #29.

File: tasks.py (hunk 7 above)
Comment (justinmayer) [reply tasks.py]: The `black` task now includes both `--check` and `--diff` options.

**Comment by justinmayer** [comment]
Follow-up commit added. Let me know what you think!

File: tasks.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
+import os
+from pathlib import Path
+from shutil import which
+
+from invoke import task
+from invoke.util import cd
+
+TOOLS = ["poetry", "pre-commit"]
+DEMO_PORT = os.environ.get("DEMO_PORT", 8000)
+DOCS_PORT = os.environ.get("DOCS_PORT", 8000)
```
Comment (Natim) [inline tasks.py]: What about using 9000 or 8888 as a default value?

File: tasks.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    if check:
+        check_flag = "--check"
+    if diff:
+        diff_flag = "--diff"
+    c.run(f"{VENV}/bin/black {check_flag} {diff_flag} kagi testproj")
+
+
+@task
+def isort(c):
+    c.run(f"{VENV}/bin/isort --recursive .isort.cfg kagi/* testproj/*")
```
Comment (Natim) [inline tasks.py]: Maybe we can also use a `--check` option for isort too.

File: tasks.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+
+@task
+def flake8(c):
+    c.run(f"{VENV}/bin/flake8 kagi testproj")
+
+
+@task
+def lint(c):
+    c.run(f"{VENV}/bin/isort -c --recursive .isort.cfg kagi/* testproj/*")
```
Comment (Natim) [inline tasks.py]:
```suggestion
    isort(c, check=True)
```

File: tasks.py [hunk 11]
```diff
@@ -0,0 +1,123 @@
+import os
+from pathlib import Path
+from shutil import which
+
+from invoke import task
+from invoke.util import cd
+
+TOOLS = ["poetry", "pre-commit"]
```
Comment (Natim) [inline tasks.py]: Should we move this on line 17 since they are related somehow?

**Comment by Natim** [comment]
Maybe we can remove the Makefile with this PR too?

**Comment by justinmayer** [comment]
Thanks for the Pathlib fix. Was just about to make a similar change. 🤝

**Pull request merged**